cranelift/codegen/src/isa/riscv64/inst/emit.rs

```diff
@@ -2821,16 +2837,30 @@ impl MachInstEmit for Inst {
                 flags,
                 ..
             } => {
-                let offset = from.get_offset_with_state(state);
-                let from_reg = allocs.next(from.get_base_register());
+                let from = from.clone().with_allocs(&mut allocs);
                 let to = allocs.next_writable(to);
 
                 // Vector Loads don't support immediate offsets, so we need to load it into a register.
-                let addr = writable_spilltmp_reg();
-                LoadConstant::U64(offset as u64)
-                    .load_constant_and_add(addr, from_reg)
-                    .into_iter()
-                    .for_each(|inst| inst.emit(&[], sink, emit_info, state));
+                let addr = match from {
+                    VecAMode::UnitStride { base } => {
+                        let base_reg = base.get_base_register();
+                        let offset = base.get_offset_with_state(state);
+
+                        // Reg+0 Offset can be directly encoded
+                        if let (Some(base_reg), 0) = (base_reg, offset) {
+                            base_reg
+                        } else {
+                            // Otherwise load the address it into a reg and load from it.
+                            let tmp = writable_spilltmp_reg();
+                            Inst::LoadAddr {
+                                rd: tmp,
+                                mem: base.clone(),
+                            }
+                            .emit(&[], sink, emit_info, state);
+                            tmp.to_reg()
```

Comment:
Should the offset be used in the `else` branch?

Reply:
`base` here has both the register and the offset inside